opr/opr.go

```diff
@@ -427,9 +429,78 @@ func (opr *OraclePriceRecord) CreateOPREntry(nonce []byte, difficulty uint64) (*
 
 	e.ChainID = hex.EncodeToString(base58.Decode(opr.OPRChainID))
 	e.ExtIDs = [][]byte{nonce, buf, {opr.Version}}
-	e.Content, err = json.Marshal(opr)
+	e.Content, err = opr.SafeMarshalJson()
 	if err != nil {
 		return nil, err
 	}
 	return e, nil
 }
+
+// SafeMarshalJson will marshal the json depending on the opr version
+func (opr *OraclePriceRecord) SafeMarshalJson() ([]byte, error) {
+	// our opr version must be set before entering this
+	if opr.Version == 0 {
+		return nil, fmt.Errorf("opr version is 0")
+	}
+
+	// This function relies on the assets, so check up front
+	if opr.Assets == nil {
+		return nil, fmt.Errorf("assets is nil, cannot marshal")
+	}
+
+	// When we marshal a version 1 opr, we need to change PEG -> PNT
+	// No opr in the code should ever have 'PNT'. We only use PNT in the marshal
+	// function, no where else.
+	if _, ok := opr.Assets["PNT"]; ok {
+		return nil, fmt.Errorf("this opr has asset 'PNT', it should have 'PEG'")
+	}
+
+	// Do the swap if on version 1
+	if opr.Version == 1 {
+		opr.Assets["PNT"] = opr.Assets["PEG"]
+		delete(opr.Assets, "PEG")
+	}
+
+	// Version 2 we do nothing
+
+	// This is a known key that will be removed by the marshal json function. It indicates
+	// to the marshaler that it was called from a safe path. This is not the cleanest method,
+	// but to override the json function, and still use the default, it would require an odd
+	// structure nesting and a lot of code changes
+	opr.Assets["version"] = float64(opr.Version)
+	data, err := json.Marshal(opr)
+	delete(opr.Assets, "version") // Should be deleted by the json.Marshal, but that can error out
+
+	// Revert version 1 changes
+	if opr.Version == 1 {
+		opr.Assets["PEG"] = opr.Assets["PNT"]
+		delete(opr.Assets, "PNT")
+	}
+
+	return data, err
+}
+
+// SafeMarshalJson will unmarshal the json depending on the opr version
+func (opr *OraclePriceRecord) SafeUnmarshalJSON(data []byte) error {
+	// our opr version must be set before entering this
+	if opr.Version == 0 {
+		return fmt.Errorf("opr version is 0")
+	}
+
+	err := json.Unmarshal(data, opr)
+	if err != nil {
+		return err
+	}
+
+	// If version 1, we need to swap PNT and PEG
```

Comment:
Can you explain why the asset needs to be swapped directly inside the OPR? I am not convinced we should alter directly the OPR retroactively from the inside... you don't delete XPD/XPT for instance

Reply:
XPD and XPT are valid v1 assets. The code internally only deals with `PEG`, there is no reference to `PNT` except in the marshaling functions. This means all v1 users when upgrading will query their `PEG` balance, not their `PNT`.

If I did not do this, everyone would have a `PNT` and `PEG` balance, and I think it would add more confusion than it is worth. We are stuck with `PNT` on chain, and that is the only place it exists.